File: python/direction.py
# Library
import os
import cv2
import math
import heapq
import numpy as np
import pandas as pd


# Main Class
class direction():
    # Initialize the class
    def __init__(self, rows: int, cols: int) -> None:
        self.toggle = False
        
        # Initialize maze size
        self.rows = rows
        self.cols = cols
        
        # Create an unknown maze
        self.maze = np.full((rows, cols), -1, dtype=np.int8)
        
        # Variables definition
        self.vision = []
        self.path = []
        
        # Path buffer heap
        self.path_tmp = []
        heapq.heapify(self.path_tmp)
    
    # Modify the path
    def path_mod(self, path: list[tuple]) -> None:
        self.path.append(path)
        
    # Modify the heap
    def path_heap(self, loc: tuple) -> tuple:
        if self.path_tmp is None:
            return None
        
        _, path = heapq.heappop(self.path_tmp)
        return (path[0] - loc[0], path[1] - loc[1])
    
    # Clear path_tmp cache
    def path_clr(self) -> tuple:
        self.path_tmp = []
        return None
    
    # Define the heuristic function
    def heuristic(self, a: tuple, b: tuple) -> float:
        return math.dist(a, b)  # Return the distance

    # Black block search
    def block_search(self, start: tuple, goal: tuple) -> list[tuple]:
        open_set = []
        heapq.heappush(open_set, (0, start))
        
        came_from = {}
        g_score = {start: 0}
        f_score = {start: self.heuristic(start, goal)}
        
        while open_set:
            _, current = heapq.heappop(open_set)
            
            if current == goal:
                # Clear path_tmp cache
                self.path_clr()
                
                idx = 0
                while current in came_from:
                    heapq.heappush(self.path_tmp, (idx, current))
                    current = came_from[current]
                    idx -= 1
                    
                # Trigger the toggle
                self.toggle = False
                
                return self.path_heap(start)
            
            for direction in [
                    (-1,  1), ( 0,  1), ( 1,  1),
                    (-1,  0),           ( 1,  0),
                    (-1, -1), ( 0, -1), ( 1, -1)
                ]:
                neighbor = (current[0] + direction[0], current[1] + direction[1])
                
                if 0 <= neighbor[0] < self.rows and 0 <= neighbor[1] < self.cols and self.maze[neighbor[0]][neighbor[1]] != 1:
                    tentative_g_score = g_score[current] + 1
                    
                    if self.maze[neighbor[0], neighbor[1]] == -1:
                        # Clear path_tmp cache
                        self.path_clr()
                        
                        idx = 0
                        while current in came_from:
                            heapq.heappush(self.path_tmp, (idx, current))
                            current = came_from[current]
                            idx -= 1
                            
                        # Trigger the toggle
                        self.toggle = False
                        
                        return self.path_heap(start)
                    
                    if neighbor not in g_score or tentative_g_score < g_score[neighbor]:
                        came_from[neighbor] = current
                        g_score[neighbor] = tentative_g_score
                        f_score[neighbor] = tentative_g_score + self.heuristic(neighbor, goal)
                        heapq.heappush(open_set, (f_score[neighbor], neighbor))
                        
        return None  # No path found

    # Find the nearest black block
    def nearest_bblock_path(self, loc: tuple) -> list[tuple]:
        block_set = []
        
        for i in range(self.rows):
            for j in range(self.cols):
                if self.maze[i, j] == -1:
                    heapq.heappush(block_set, (math.dist((i, j), loc), (i, j)))
                if self.maze[i, j] == 2:
                    direct = self.destination(start=loc, goal=(i, j))
                    if direct is not None:
                        return direct
                    
        while block_set:
            _, block_loc = heapq.heappop(block_set)
            
            direct = self.block_search(start=loc, goal=block_loc)
            
            if direct is not None:
                block_set = []
                return direct
        
        return None # Path not found

    # Check the destination
    def destination(self, start: tuple, goal: tuple) -> tuple:
        open_set = []
        heapq.heappush(open_set, (0, start))
        direction = (0, 0)
        
        came_from = {}
        g_score = {start: 0}
        f_score = {start: self.heuristic(start, goal)}
        
        while open_set:
            _, current = heapq.heappop(open_set)
            
            if current == goal:
                # Clear path_tmp cache
                self.path_clr()
                
                idx = 0
                while current in came_from:
                    heapq.heappush(self.path_tmp, (idx, current))
                    current = came_from[current]
                    idx -= 1
                    
                # Trigger the toggle
                self.toggle = True
                
                return self.path_heap(start)
            
            for direction in [
                    (-1,  1), ( 0,  1), ( 1,  1),
                    (-1,  0),           ( 1,  0),
                    (-1, -1), ( 0, -1), ( 1, -1)
                ]:
                neighbor = (current[0] + direction[0], current[1] + direction[1])
                
                if 0 <= neighbor[0] < self.rows and \
                    0 <= neighbor[1] < self.cols and \
                    (self.maze[neighbor[0]][neighbor[1]] == 0 or \
                    self.maze[neighbor[0]][neighbor[1]] == 2):
                    tentative_g_score = g_score[current] + 1
                    
                    if neighbor not in g_score or tentative_g_score < g_score[neighbor]:
                        came_from[neighbor] = current
                        g_score[neighbor] = tentative_g_score
                        f_score[neighbor] = tentative_g_score + self.heuristic(neighbor, goal)
                        heapq.heappush(open_set, (f_score[neighbor], neighbor))
        
        return None  # No path found
    
    # Visualize the nearby blocks
    def visualize(self, x_buffer: np.array, y_buffer: np.array, v_buffer: np.array, length: int) -> tuple:
        dest = None
        
        # Vision info load
        for i in range(length):
            # Read vision to the maze
            x = x_buffer[i]
            y = y_buffer[i]
            self.maze[x, y] = v_buffer[i]
            
            # Check destination
            if v_buffer[i] == 2:
                dest = (x, y)
            
            # Initialize vision
            self.vision.append((x, y))
            
        return dest
    
    # Return the intended direcion
    def direct(self, loc: tuple) -> tuple:
        # Variables init
        current = loc
        maxBlock = 0
        move_dir = (0, 0)
        
        # Direction loop
        for direction in [
                    (-1,  1), ( 0,  1), ( 1,  1),
                    (-1,  0),           ( 1,  0),
                    (-1, -1), ( 0, -1), ( 1, -1)
                ]:
            
            # Calculate the number of the black block to be discovered
            neighbor = (current[0] + direction[0], current[1] + direction[1])
            if not (0 <= neighbor[0] < self.rows and 0 <= neighbor[1] <self.cols) or self.maze[neighbor[0], neighbor[1]] != 0: continue
            
            block = 0
            for coordinate in self.vision:
                    if 0 <= coordinate[0] + direction[0] < self.rows and \
                       0 <= coordinate[1] + direction[1] <self.cols and \
                       self.maze[coordinate[0] + direction[0], coordinate[1] + direction[1]] == -1:
                        block += 1
            
            if block > maxBlock:
                maxBlock = block
                move_dir = direction
        
        return move_dir
    
    def final_check(self, start: tuple) -> tuple:
        for i in range(self.rows):
            for j in range(self.cols):
                if self.maze[i, j] == 2:
                    direct = self.destination(start=start, goal=(i, j))
                    return direct
        
        return None  # Goal not found

dir = None

# Initialize the map
def init(width: int, height: int) -> None:
    global dir
    dir = direction(cols=width, rows=height)
    return None

# Result
def result(x: int, y: int, x_buffer: np.array, y_buffer: np.array, v_buffer: np.array, length: int) -> tuple:
    global dir
    
    assert dir is not None, 'ERROR: Class not initialized!'
    
    direct = (0, 0)
    
    # Add current location to path
    dir.path_mod(path=(x, y))
    
    # Vision init
    dest = dir.visualize(x_buffer=x_buffer, y_buffer=y_buffer, v_buffer=v_buffer, length=length)
    
    # Destination rush
    if dir.toggle == True:
        # print((x +direct[0], y +direct[1]))
        direct = dir.path_heap(loc=(x, y))
        assert direct is not None, 'ERROR: \'direct\' is \'NoneType\'!'
        return direct
    
    # Destination check
    if isinstance(dest, tuple):
        # print(dest)
        direct = dir.destination(start=(x, y), goal=dest)
        if direct is not None:
            return direct
        
    # Find the direction
    direct = dir.nearest_bblock_path(loc=(x, y))
    if direct is not None:
        return direct
    
    direct = dir.final_check(start=(x, y))
    assert direct is not None, 'ERROR: \'direct\' is \'NoneType\'!'
    return direct
        

maze = np.array([
    [0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1],
    [0, 1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1],
    [0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 0, 1],
    [1, 1, 1, 1, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 0, 1],
    [0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 1, 1, 0, 1],
    [0, 1, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 1],
    [0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0],
    [0, 1, 1, 1, 0, 1, 0, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1],
    [0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1],
    [0, 1, 0, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 0, 1, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1],
    [0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1],
    [1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 0, 1, 1, 1, 0, 1, 0, 1],
    [0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1],
    [0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 1],
    [0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0],
    [1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1],
    [0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1],
    [0, 1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1],
    [0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1],
    [1, 1, 0, 1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1],
    [0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1],
    [0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 1, 1, 1, 0, 1, 0, 2]], dtype=np.int8)

inverted_array = np.array([[not element for element in row] for row in maze])
image_array = (inverted_array * 255).astype(np.uint8)
image = cv2.cvtColor(image_array, cv2.COLOR_GRAY2BGR)
cv2.imwrite("map.png", image)

rows = 22
cols = 32
init(width=32, height=22)

def plot_res(maze=maze, path=None, filename='tmp.png') -> None:
    
    res = maze.copy()
    row = rows
    col = cols
    
    rgb = np.full((rows, cols, 3), fill_value=0, dtype=np.int8)
    
    for i in range(row):
        for  j in range(col):
            if res[i, j] == 1:
                rgb[i, j] = np.array([0, 0, 0])
                continue
            
            if res[i, j] == 0:
                rgb[i, j] = np.array([255, 255, 255])
                continue
            
            if res[i, j] == -1:
                rgb[i ,j] = np.array([0, 0, 0])
 
    if path:
        for step in path:
            rgb[step[0], step[1]] += np.array([50, 50, 0])
                
    image = np.array(rgb).astype(np.uint8)
    # image = cv2.cvtColor(image_array, cv2.COLOR_RGB2BGR)
    cv2.imwrite(filename, image)
    
start = (0, 0)
goal = (21, 31)
current = start
radius = 2
path = []
print(np.size(maze))

while current != goal:
    path.append(current)
    
    x_range = (max(0, current[0]-radius), min(rows, current[0]+radius+1))
    y_range = (max(0, current[1]-radius), min(cols, current[1]+radius+1))
    
    x_buffer = []
    y_buffer = []
    v_buffer = []
    
    length = 0
    for i in range(x_range[0], x_range[1]):        
        for j in range(y_range[0], y_range[1]):
            x_buffer.append(i)
            y_buffer.append(j)
            v_buffer.append(maze[i, j])
            length += 1
    
    direct = result(x=current[0], y=current[1], x_buffer=x_buffer, y_buffer=y_buffer, v_buffer=v_buffer, length=length)
    
    plot_res(maze=dir.maze, path=dir.path)
    
    current = (current[0] + direct[0], current[1] + direct[1])
    
print(path)

File: src/direction.cpp
// direction.cpp
//
#include "../include/direction.h"

// Initialize the class
Direction::Direction(int row, int column) : rows(row), cols(column), toggle(false) {
    // Initialize a black maze
    maze.resize(rows, std::vector<int8_t>(cols, -1));
}

// Append sub path to the path
void Direction::path_mod(const std::vector<std::tuple<int, int>>& sub_path) {
    std::move(sub_path.begin(), sub_path.end(), std::back_inserter(path));
}

// Return the direction from path
std::tuple<int, int> Direction::path_pop(const std::tuple<int, int>& loc) {
    if (path_tmp.empty()) return std::make_tuple(0, 0);

    std::tuple<int, int> ret_path = std::get<1>(path_tmp.top());
    path_tmp.pop();

    return std::make_tuple(std::get<0>(ret_path) - std::get<0>(loc), std::get<1>(ret_path) - std::get<1>(loc));
}

// Clear the path
void Direction::path_clr() {
    while (!path_tmp.empty()) path_tmp.pop();
}

// Calculate the distance of two blocks
float Direction::heuristic(const std::tuple<int, int>& a, const std::tuple<int, int>& b) {
    return (float)std::sqrt(std::pow(std::get<0>(a) - std::get<0>(b), 2) + std::pow(std::get<1>(a) - std::get<1>(b), 2));
}

// Black block search
std::tuple<int, int> Direction::block_search(const std::tuple<int, int>& start, const std::tuple<int, int>& goal) {
    std::priority_queue<
        std::tuple<float, std::tuple<int, int>>,
        std::vector<std::tuple<float, std::tuple<int, int>>>,
        std::greater<>> open_set;

    open_set.push(std::make_tuple(0.0f, start));

    std::unordered_map<std::tuple<int, int>, std::tuple<int, int>, tuple_hash> came_from;
    std::unordered_map<std::tuple<int, int>, int, tuple_hash> g_score;
    g_score[start] = 0;
    std::unordered_map<std::tuple<int, int>, float, tuple_hash> f_score;
    f_score[start] = heuristic(start, goal);

    while (!open_set.empty()) {
        // Debug
        /*std::cout << "open_set: " << std::endl;
        Direction::print_priority_queue(open_set);*/

        std::tuple<int, int> current = std::get<1>(open_set.top());
        open_set.pop();

        if (current == goal) {
            // Clear path_tmp cache
            path_clr();
            
            // Debug
            /*std::cout << "Block Search 1: " << std::get<0>(current) << ", " << std::get<1>(current) << std::endl;*/

            int idx = 0;
            while (came_from.find(current) != came_from.end()) {
                path_tmp.emplace(idx, current);
                current = came_from[current];
                idx -= 1;
            }
            
            // Trigger the toggle
            toggle = false;
            
            return path_pop(start);
        }

        for (auto dir : dirs) {
            std::tuple<int, int> neighbor = {std::get<0>(current) + std::get<0>(dir), std::get<1>(current) + std::get<1>(dir)};

            // Check the boundry
            if (std::get<0>(neighbor) >= 0 && std::get<0>(neighbor) < rows &&
                std::get<1>(neighbor) >= 0 && std::get<1>(neighbor) < cols &&
                maze[std::get<0>(neighbor)][std::get<1>(neighbor)] != 1) {
                
                /* Check the corner wall
                * if the two corners are blocked, then continue to next direction
                */
                if (std::abs(std::get<0>(dir)) + std::abs(std::get<1>(dir)) == 2) {  // Get the diagonal position, that's when the manhatten distance is 2
                    std::tuple<int, int> corner1 = { std::get<0>(current) + std::get<0>(dir), std::get<1>(current) };
                    std::tuple<int, int> corner2 = { std::get<0>(current), std::get<1>(current) + std::get<1>(dir) };

                    // if both of the corner are not path, then continue
                    if (maze[std::get<0>(corner1)][std::get<1>(corner1)] != 0 &&
                        maze[std::get<0>(corner2)][std::get<1>(corner2)] != 0) {
                        // Debug
                        /*std::cout << "Corner Check: (" << std::get<0>(neighbor) << ", " << std::get<1>(neighbor)
                            << ") = " << static_cast<int>(maze[std::get<0>(neighbor)][std::get<1>(neighbor)]) << std::endl;*/

                        continue;
                    }
                }

                // Main
                int tentative_g_score = g_score[current] + 1;
                
                // Debug
                /*std::cout << "Block Search 3: " << std::get<0>(neighbor) << ", " << std::get<1>(neighbor) 
                    << ", " << static_cast<int>(maze[std::get<0>(neighbor)][std::get<1>(neighbor)]) << std::endl;*/

                if (maze[std::get<0>(neighbor)][std::get<1>(neighbor)] == -1) {
                    // Clear path_tmp cache
                    path_clr();
                    
                    // Debug
                    /*std::cout << "Block Search 2: " << std::get<0>(neighbor) << ", " << std::get<1>(neighbor) << std::endl;*/

                    int idx = 0;
                    while (came_from.find(current) != came_from.end()) {
                        path_tmp.emplace(idx, current);
                        current = came_from[current];
                        idx -= 1;
                    }
                    
                    // Trigger the toggle
                    toggle = false;
                    
                    return path_pop(start);
                }
                
                if (g_score.find(neighbor) == g_score.end() || tentative_g_score < g_score[neighbor]) {
                    came_from[neighbor] = current;
                    g_score[neighbor] = tentative_g_score;
                    f_score[neighbor] = tentative_g_score + heuristic(neighbor, goal);
                    open_set.emplace(f_score[neighbor], neighbor);
                }
            }
        }
    }

    // Path not found
    return std::make_tuple(0, 0);
}

// Destination rush
std::tuple<int, int> Direction::destination(const std::tuple<int, int>& start, const std::tuple<int, int>& goal) {
    std::priority_queue<
        std::tuple<float, std::tuple<int, int>>,
        std::vector<std::tuple<float, std::tuple<int, int>>>,
        std::greater<>> open_set;

    open_set.push(std::make_tuple(0.0f, start));

    std::unordered_map<std::tuple<int, int>, std::tuple<int, int>, tuple_hash> came_from;
    std::unordered_map<std::tuple<int, int>, int, tuple_hash> g_score;
    g_score[start] = 0;
    std::unordered_map<std::tuple<int, int>, float, tuple_hash> f_score;
    f_score[start] = heuristic(start, goal);

    while (!open_set.empty()) {
        std::tuple<int, int> current = std::get<1>(open_set.top());
        open_set.pop();

        if (current == goal) {
            // Clear path_tmp cache
            path_clr();
            
            int idx = 0;
            while (came_from.find(current) != came_from.end()) {
                path_tmp.emplace(idx, current);
                current = came_from[current];
                idx -= 1;
            }
            
            // Trigger the toggle
            toggle = false;
            
            return path_pop(start);
        }

        for (auto dir : dirs) {
            std::tuple<int, int> neighbor = {std::get<0>(current) + std::get<0>(dir), std::get<1>(current) + std::get<1>(dir)};

            // Check the boundry
            if (std::get<0>(neighbor) >= 0 && std::get<0>(neighbor) < rows &&
                std::get<1>(neighbor) >= 0 && std::get<1>(neighbor) < cols &&
                (maze[std::get<0>(neighbor)][std::get<1>(neighbor)] == 0 ||
                 maze[std::get<0>(neighbor)][std::get<1>(neighbor)] == 2)) {
                
                /* Check the corner wall
                * if the two corners are blocked, then continue to next direction
                */
                if (std::abs(std::get<0>(dir)) + std::abs(std::get<1>(dir)) == 2) {  // Get the diagonal position, that's when the manhatten distance is 2
                    std::tuple<int, int> corner1 = { std::get<0>(current) + std::get<0>(dir), std::get<1>(current) };
                    std::tuple<int, int> corner2 = { std::get<0>(current), std::get<1>(current) + std::get<1>(dir) };

                    // if both of the corner are not path, then continue
                    if (maze[std::get<0>(corner1)][std::get<1>(corner1)] != 0 &&
                        maze[std::get<0>(corner2)][std::get<1>(corner2)] != 0) {
                        // Debug
                        /*std::cout << "Corner Check: (" << std::get<0>(neighbor) << ", " << std::get<1>(neighbor)
                            << ") = " << static_cast<int>(maze[std::get<0>(neighbor)][std::get<1>(neighbor)]) << std::endl;*/

                        continue;
                    }
                }

                // Main
                int tentative_g_score = g_score[current] + 1;
                
                if (maze[std::get<0>(neighbor)][std::get<1>(neighbor)] == -1) {
                    // Clear path_tmp cache
                    path_clr();
                    
                    int idx = 0;
                    while (came_from.find(current) != came_from.end()) {
                        path_tmp.emplace(idx, current);
                        current = came_from[current];
                        idx -= 1;
                    }
                    
                    // Trigger the toggle
                    toggle = false;
                    
                    return path_pop(start);
                }
                
                if (g_score.find(neighbor) == g_score.end() || tentative_g_score < g_score[neighbor]) {
                    came_from[neighbor] = current;
                    g_score[neighbor] = tentative_g_score;
                    f_score[neighbor] = tentative_g_score + heuristic(neighbor, goal);
                    open_set.emplace(f_score[neighbor], neighbor);
                }
            }
        }
    }

    // Path not found
    return std::make_tuple(0, 0);
}

// Find the nearest black block
std::tuple<int, int> Direction::nearest_bblock_path(const std::tuple<int, int>& loc) {
    std::priority_queue<std::tuple<float, std::tuple<int, int>>, 
                        std::vector<std::tuple<float, std::tuple<int, int>>>, 
                        std::greater<>> block_set;

    for (int i = 0; i < rows; ++i) {
        for (int j = 0; j < cols; ++j) {
            if (maze[i][j] == -1) {
                block_set.emplace(sqrt(pow(i - std::get<0>(loc), 2) + pow(j - std::get<1>(loc), 2)), std::make_tuple(i, j));
            }
            if (maze[i][j] == 2) {
                auto direct = destination(loc, std::make_tuple(i, j));
                if (direct != std::make_tuple(0, 0)) {
                    // Debug
                    // std::cout << "Direct ret: " << std::get<0>(direct) << ", " << std::get<1>(direct) << std::endl;

                    return direct;
                }
            }
        }
    }


    std::tuple<int, int> block_loc;
    std::tuple<float, std::tuple<int, int>> tmp;
    while (!block_set.empty()) {
        std::tuple<int, int> block_loc = std::get<1>(block_set.top());
        block_set.pop();

        // Debug
        //std::cout << "Block loc: " << std::get<0>(block_loc) << ", " << std::get<1>(block_loc) << std::endl;

        auto direct = block_search(loc, block_loc);

        if (direct != std::make_tuple(0, 0)) {
            while (!block_set.empty()) block_set.pop();  // Clear the block set
            return direct;
        }
    }

    // Path not found
    return std::make_tuple(0, 0);
}

// Vision
std::tuple<int, int> Direction::visualize(const std::vector<int>& x_buffer, const std::vector<int>& y_buffer, const std::vector<int8_t>& v_buffer, int length) {
    std::tuple<int, int> dest = {0, 0};

    for (int i = 0; i < length; ++i) {
        // Read vision to the maze
        int x = x_buffer[i];
        int y = y_buffer[i];

        /*std::cout << "Check Point 3: " << x << ", " << y << "   " << static_cast<int>(v_buffer[i]) << std::endl;*/

        maze[x][y] = v_buffer[i];

        // Check destination
        if (v_buffer[i] == 2) {
            dest = std::make_tuple(x, y);
        }

        // Initialize vision
        vision.emplace_back(x, y);
    }

    return dest;
}

// Find the direction which can score the most black blocks
std::tuple<int, int> Direction::direct(std::tuple<int, int> loc) {
    // Variables init
    std::tuple<int, int> current = loc;
    int maxBlock = 0;
    std::tuple<int, int> move_dir = {0, 0};

    // Direction loop
    for (const auto& dir : dirs) {
        // Calculate the number of the black block to be discovered
        std::tuple<int, int> neighbor = {std::get<0>(current) + std::get<0>(dir), std::get<1>(current) + std::get<1>(dir)};

        if (!(0 <= std::get<0>(neighbor) && std::get<0>(neighbor) < rows 
            && 0 <= std::get<1>(neighbor) && std::get<1>(neighbor) < cols) 
            || maze[std::get<0>(neighbor)][std::get<1>(neighbor)] != 0) {
            continue;
        }

        int block = 0;
        for (const auto& coordinate : vision) {
            if (0 <= std::get<0>(coordinate) + std::get<0>(dir) && std::get<0>(coordinate) + std::get<0>(dir) < rows 
                && 0 <= std::get<1>(coordinate) + std::get<1>(dir) && std::get<1>(coordinate) + std::get<1>(dir) < cols 
                && maze[std::get<0>(coordinate) + std::get<0>(dir)][std::get<1>(coordinate) + std::get<1>(dir)] == -1) {
                block += 1;
            }
        }

        if (block > maxBlock) {
            maxBlock = block;
            move_dir = dir;
        }
    }

    return move_dir;
}

std::tuple<int, int> Direction::final_check(std::tuple<int, int> start) {
        for (int i = 0; i < rows; ++i) {
            for (int j = 0; j < cols; ++j) {
                if (maze[i][j] == 2) {
                    auto direct = destination(start, std::make_tuple(i, j));
                    if (direct != std::make_tuple(0, 0)) {
                        return direct;
                    }
                }
            }
        }

        return std::make_tuple(0, 0);  // Goal not found
}

// Return type redef
int* Direction::redef(const std::tuple<int, int>& dir) {
    int* ret = (int*)malloc(2 * sizeof(int));
    if (ret) {
        ret[0] = std::get<0>(dir);
        ret[1] = std::get<1>(dir);
        // Debug
        // std::cout << "ret: " << ret[0] << ", " << ret[1] << std::endl;
    }
    else {
        std::cerr << "Error: Null pointer dereferenced." << std::endl;
    }

    return ret;
}

// DEBUG
// 
// Function to print the contents of the priority queue
void Direction::print_priority_queue(std::priority_queue<
    std::tuple<float, std::tuple<int, int>>,
    std::vector<std::tuple<float, std::tuple<int, int>>>,
    std::greater<>> open_set) {
    // Create a vector to hold the elements of the priority queue
    std::vector<std::tuple<float, std::tuple<int, int>>> elements;

    // Copy the elements of the priority queue to the vector
    while (!open_set.empty()) {
        elements.push_back(open_set.top());
        open_set.pop();
    }

    // Print the elements
    for (const auto& element : elements) {
        float priority = std::get<0>(element);
        auto coords = std::get<1>(element);
        int x = std::get<0>(coords);
        int y = std::get<1>(coords);
        std::cout << "Priority: " << priority << ", Coordinates: (" << x << ", " << y << ")" << std::endl;
    }
}

// Print maze
void Direction::draw_maze() {
    int i, j;
    for (i = 0; i < rows; i++) {
        for (j = 0; j < cols; j++) {
            std::cout << std::setw(2) << static_cast<int>(maze[i][j]) << " ";
        }
        std::cout << std::endl;
    }
}